cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/cloudstorage/coordinated/CoordinatedCloudStorageDataTransferApi.java

```diff
@@ -0,0 +1,236 @@
+                                                                                qualifiedTableName.table(),
+                                                                                jobInfo.getRestoreJobId(),
+                                                                                createSliceRequestPayload);
+        RetryPolicy retryPolicy = new CloudStorageDataTransferApiImpl.ExecutorCreateSliceRetryPolicy(sidecarClient);
+        RequestContext requestContext = sidecarClient.requestBuilder().retryPolicy(retryPolicy).request(request).build();
+        try
+        {
+            sidecarClient.executeRequestAsync(requestContext).get();
+        }
+        catch (Exception exception)
+        {
+            handleInterruption(exception);
+            throw new SidecarApiCallException("Failed to create restore job slice. " +
+                                              "restoreJobId=" + jobInfo.getRestoreJobId() +
+                                              " sliceId=" + createSliceRequestPayload.sliceId(),
+                                              exception);
+        }
+    }
+
+    private void restoreJobProgressInternal(String clusterId,
+                                            CloudStorageDataTransferApiImpl dataTransferApi,
+                                            RestoreJobProgressFetchPolicy fetchPolicy,
+                                            BiConsumer<String, RestoreJobProgressResponsePayload> progressHandler)
+    {
+        JobInfo jobInfo = dataTransferApi.jobInfo();
+        QualifiedTableName qualifiedTableName = jobInfo.qualifiedTableName();
+        UUID restoreJobId = jobInfo.getRestoreJobId();
+        RestoreJobProgressRequestParams requestParams = new RestoreJobProgressRequestParams(qualifiedTableName.keyspace(),
+                                                                                            qualifiedTableName.table(),
+                                                                                            restoreJobId,
+                                                                                            fetchPolicy);
+        RestoreJobProgressResponsePayload jobProgress = restoreJobProgress(dataTransferApi, requestParams);
+        progressHandler.accept(clusterId, jobProgress);
+    }
+
+    private void forEachInternal(Consumer<CloudStorageDataTransferApiImpl> dataTransferApiConsumer)
+    {
+        Predicate<String> allowAllClusters = id -> false;
+        forEachInternal(allowAllClusters, (ignored, api) -> dataTransferApiConsumer.accept(api));
+    }
+
+    // run dataTransferApiAction on the clusters that are not filtered by clusterIdFilter
+    private void forEachInternal(Predicate<String> clusterIdFilter,
+                                 BiConsumer<String, CloudStorageDataTransferApiImpl> dataTransferApiAction)
+    {
+        dataTransferApis.forEach((clusterId, dataTransferApi) -> {
+            if (clusterIdFilter.test(clusterId))
```

Comment:
Usually `filter`-type functions _include_ the items that _match_ the predicate, not the other way around. Can you swap this so that the predicate is used in a similar way to other filters like `Stream.filter`? In fact, there are two places where this is used, and each seems to think that it is a different behavior.

This one returns true for every one (which means it never runs):
```java
   @Override
    public void forEach(BiConsumer<String, CloudStorageDataTransferApiImpl> action)
    {
        forEachInternal(ignored -> true, action);
    }
```

This one returns false (so will run for all clusters):
```java
    private void forEachInternal(Consumer<CloudStorageDataTransferApiImpl> dataTransferApiConsumer)
    {
        Predicate<String> allowAllClusters = id -> false;
        forEachInternal(allowAllClusters, (ignored, api) -> dataTransferApiConsumer.accept(api));
    }
```

Reply:
Yeah. Good catch. 
They are cancelling each other. I have updated the code.